net/api/test_server.cc.h

```diff
@@ -9,17 +9,216 @@
 #ifndef BRICKS_NET_API_TEST_SERVER_CC_H
 #define BRICKS_NET_API_TEST_SERVER_CC_H
 
-#include "../../dflags/dflags.h"
+#include <thread>
 
+#include "api.h"
+#include "../url/url.h"
 #include "../http/test_server.cc.h"
 
+#include "../../strings/printf.h"
+
 namespace bricks {
 namespace net {
 namespace api {
 namespace test {
 
-struct TestHTTPServer_APIImpl : bricks::net::http::test::TestHTTPServer_HTTPImpl {
-  // TODO(dkorolev): Real API implementation goes here.
+enum class Method { UNINITIALIZED, GET, POST };
+
+struct Request {
+  const url::URL& url;
+  HTTPServerConnection& connection;
+  const HTTPReceivedMessage& message;
+};
+
+struct Routes {
+  struct Helper {
+    struct Impl {
+      explicit Impl(Routes* routes) : routes_(routes) {}
+      ~Impl() {
+        assert(method_ != Method::UNINITIALIZED);
+        assert(!path_.empty());
+        assert(handler_);
+        routes_->AddRoute(method_, path_, handler_);
+      }
+      void SetHandler(std::function<void(Request&&)>&& f) {
+        assert(!handler_);
+        handler_ = f;
+      }
+      void AddConstraint(Method method) {
+        assert(method != Method::UNINITIALIZED);
+        assert(method_ == Method::UNINITIALIZED);
+        method_ = method;
+      }
+      void AddConstraint(const std::string& path) {
+        assert(!path.empty());
+        assert(path_.empty());
+        path_ = path;
+      }
+
+      Routes* routes_;
+      std::function<void(Request&&)> handler_;
+
+      Method method_ = Method::UNINITIALIZED;
+      std::string path_ = "";
+
+      Impl() = delete;
+      Impl(const Impl&) = delete;
+      Impl(Impl&&) = delete;
+      void operator=(const Impl&) = delete;
+      void operator=(Impl&&) = delete;
+    };
+
+    Helper(Routes* routes) : impl_(new Impl(routes)) {}
+    Helper(Helper&& rhs) { impl_.swap(rhs.impl_); }
+    void operator=(std::function<void(Request&&)>&& f) {
+      assert(impl_);
+      impl_->SetHandler(std::forward<std::function<void(Request && )>>(f));
+    }
+    template <typename T>
+    Helper& operator[](T&& constraint) {
+      assert(impl_);
+      impl_->AddConstraint(std::forward<T>(constraint));
+      return *this;
+    }
+
+    std::unique_ptr<Impl> impl_;
```

Comment:
To deal with move-construction seamlessly.

One ```routes[foo][bar][baz] = meh;``` should result in exactly one ```Impl``` object created, populated with several ```AddConstraint()``` and one ```SetHandler()```, and then call ```routes_->AddRoute()``` once for the whole statement.

Returning from a top-level ```operator[]``` results in  top-level object copied; thus, I made the instance of ```Helper``` to be copied, while the ```Impl``` one stays one.